### PR TITLE
Issue #244: Add before_session_start to include waiting for session locks

### LINKED ISSUE #244: Excimer isn't correctly including the time waiting for a session lock
This is hard, because we are relying on the _after_config hook and this delay is in the session manager before this hook is called.

It gives misleading numbers on pages where there could be say a 10 second delay waiting and then 1 second of processing, and excimer shows it as just one second. 

I think at minimum lets try and infer the time the script started and just show this in the profile metadata even if it isn't in the flame graph.

The tideways profiler is hard coded to allow it to kick off much earlier:
https://github.com/moodle/moodle/blob/master/lib/setup.php#L442

A proper solution might be:
* [x] land this tracker: https://tracker.moodle.org/browse/MDL-75014
* [ ] document backporting that + patch in ci for this plugin
* [ ] start excimer in that callback

Good test is using this:
https://master.localhost/admin/tool/sessionbreaker/

### PR DESCRIPTION
**Pull request by petersistrom**
Resolves #244

### CODE REVIEW

File: lib.php
Comment (brendanheywood) [inline lib.php]: because you have refactored it I would call this the same here manager::create_if_needed();

Comment (brendanheywood) [reply lib.php]: I'd also move them side by side, or better yet in the order they will be executed so this one first

File: classes/web_processor.php
Comment (brendanheywood) [inline classes/web_processor.php]: This seems wrong? Why would we not want to save the record if we don't have a user yet? Is this related to this patch?

Comment (petersistrom) [reply classes/web_processor.php]: To test this I was putting a sleep() on one page to lock the session and then trying to load another at the same time. I was getting this error when the second page loaded, as it was trying to save a null user. 

image

File: lib.php
Comment (petersistrom) [reply lib.php]: Updated

File: classes/web_processor.php
Comment (jaypha) [reply classes/web_processor.php]: This could be because it is trying to save user info before the session has started, therefore no user information is available.

**Comment by jaypha** [comment]
Creating the manager as is before the session starts will cause problems as not all information is available yet (e.g. $USER). The manager needs to start the profiler and timer while in `tool_excimer_before_session_start` but not flesh out the details until `tool_excimer_after_config`.

File: classes/web_processor.php
Comment (brendanheywood) [reply classes/web_processor.php]: As an explicit test scenario I'd like you to test a page which is holding the session lock open in one page for say 10 seconds, and then we are profiling a second page which is blocked by the first page. If you set the numbers really low so that it is saving partial profiles every 1 seconds, then I would expect the profile for page B to exist without any user data after 1 second and to show all of its time spent in the session waiting for a lock. Then after 10 seconds the lock will be released, the user will be logged in, and the profile metadata will be updated to include the user that we now know who they are

**Comment by jaypha** [comment]
I'm fine with this. Not sure if it should be merged now, or wait until the tracker lands and is backported.

**Comment by brendanheywood** [comment]
Lets wait just a little, I'm hopeful the tracker will land soon. It isn't in this weeks batch but should be in the next week

File: lib.php
Comment (brendanheywood) [inline lib.php:51]: Why is the logic for what _session_start is doing different to what _after config is doing? I would assume they should be identical

**Comment by jaypha** [comment]
The trackers have been backported.

**Review by keevan — CHANGES_REQUESTED** [review]
As mentioned recently the tracker https://tracker.moodle.org/browse/MDL-75014 has landed and so this PR will also need to include the patch for the tracker.

Once that's done and remaining review items addressed it can be merged.

File: README.md
Comment (keevan) [inline README.md]: This is good but I don't believe this is required here. It might be useful in the wiki / reusable workflows repository if not already.

File: lib.php
Comment (petersistrom) [reply lib.php:51]: This logic had to be separated out to start the timer `tool_excimer_before_session_start` and then flesh the details in `tool_excimer_after_config` as explained here https://github.com/catalyst/moodle-tool_excimer/pull/262#issuecomment-1237780878

**Review by keevan — APPROVED** [review]

**Pull request merged**
